incubator/hnc/pkg/controllers/object_controller_new.go

```diff
@@ -17,16 +17,37 @@ package controllers
 
 import (
 	"context"
+	"reflect"
 
 	"github.com/go-logr/logr"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/metadata"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/object"
+	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
 
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
 	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/forest"
 )
 
+// action is the action to take after Reconcile syncs with the in-memory forest.
+// This is introduced to consolidate calls with forest lock.
+type action int
+
+const (
+	// Start with an “unknown” to be sure of enum’s initialization.
+	unknown action = iota
+	deleting
```

Comment:
Nit suggestion: this might easier to read as a command (delete/write/ignore) rather than a state.

Reply:
sg except "delete' collides with builtin function, so I will use "remove" instead?

Reply:
👍